ts/components/messages/PaymentButton.tsx

```diff
@@ -166,83 +74,125 @@ const getButtonText = (
   });
 };
 
+/**
+ * A component to render the button related to the payment
+ * paired with a message.
+ */
 class PaymentButton extends React.PureComponent<Props> {
-  // tslint:disable-next-line: cognitive-complexity
-  public render() {
+  private hideIcon =
+    isUnexpirable(this.props.messagePaymentExpirationInfo) ||
+    isExpired(this.props.messagePaymentExpirationInfo) ||
+    !this.props.small;
+
+  private navigateToMessageDetail = () => {
+    this.props.navigateToMessageDetail(this.props.message.id);
+  };
+
+  private handleOnPress = () => {
     const {
-      paid,
       messagePaymentExpirationInfo,
-      small,
+      service,
+      paid,
       disabled,
-      onPress
+      message
     } = this.props;
+    const expired = isExpired(this.props.messagePaymentExpirationInfo);
+    const amount = getAmountFromPaymentAmount(
+      messagePaymentExpirationInfo.amount
+    );
 
-    const appliedStyles = getExpirationStyles(messagePaymentExpirationInfo);
+    const rptId = fromNullable(service).chain(_ =>
+      getRptIdFromNoticeNumber(
+        _.organization_fiscal_code,
+        messagePaymentExpirationInfo.noticeNumber
+      )
+    );
 
-    const isUnexpirable = messagePaymentExpirationInfo.kind === "UNEXPIRABLE";
-    const isExpired =
-      messagePaymentExpirationInfo.kind === "EXPIRABLE" &&
-      messagePaymentExpirationInfo.expireStatus === "EXPIRED";
-    const hideIcon = isUnexpirable || isExpired || !small;
+    if (expired) {
+      this.navigateToMessageDetail();
```

Comment:
```suggestion
      this.navigateToMessageDetail();
      return
```
I guess we should block the execution here if we are moving to another screen

Reply:
updated within https://github.com/pagopa/io-app/pull/1771/commits/6f66f44f15e5aedf8bd3a9fdf859a23617897e66